src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,465 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    ///  Represents a JSON object.
+    /// </summary>
+    public sealed class JsonObject : JsonNode, IEnumerable<KeyValuePair<string, JsonNode>>
+    {
+        private readonly Dictionary<string, JsonNode> _dictionary;
+        private readonly DuplicatePropertyNameHandling _duplicatePropertyNameHandling;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing the empty object.
+        /// </summary>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided manner of handling duplicates does not exist.
+        /// </exception>
+        public JsonObject(DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+        {
+            if(!Enum.IsDefined(typeof(DuplicatePropertyNameHandling), duplicatePropertyNameHandling))
+            {
+                throw new ArgumentException(SR.InvalidDuplicatePropertyNameHandling);
+            }
+
+            _dictionary = new Dictionary<string, JsonNode>();
+            _duplicatePropertyNameHandling = duplicatePropertyNameHandling;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing provided set of JSON properties.
+        /// </summary>
+        /// <param name="jsonProperties">>Properties to represent as a JSON object.</param>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided collection contains duplicates if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public JsonObject(IEnumerable<KeyValuePair<string, JsonNode>> jsonProperties, DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+            : this(duplicatePropertyNameHandling)
+            => AddRange(jsonProperties);
+
+        /// <summary>
+        ///   Gets or sets the value of the specified property.
+        /// </summary>
+        /// <param name="propertyName">The property name of the value to get or set.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   Provided property name is null.
+        /// </exception>
+        public JsonNode this[string propertyName]
+        {
+            get => propertyName != null ? GetPropertyValue(propertyName) : throw new ArgumentNullException(nameof(propertyName));
+            set => _dictionary[propertyName] = value ?? throw new ArgumentNullException(nameof(propertyName));
```

Comment:
Shouldn't we also validate here according to the DuplicatePropertyNameHandling setting?

Reply:
Don't we want the indexer to always change the value?

Reply:
What happens if the passed in property name doesn't exist in the dictionary? throw KeyNotFoundException?

Reply:
Inner dictionary call will create a new one with provided name and value and I think that's the behavior what we also want.